module/module.go

```diff
@@ -0,0 +1,422 @@
+
+// ToProto converts the HandlerMap to a protobuf representation.
+func (h HandlerMap) ToProto() *pb.HandlerMap {
+	pMap := &pb.HandlerMap{}
+	for s, models := range h {
+		subtype := &robotpb.ResourceRPCSubtype{
+			Subtype: protoutils.ResourceNameToProto(resource.Name{
+				Subtype: s.Subtype,
+				Name:    "",
+			}),
+			ProtoService: s.SvcName,
+		}
+
+		handler := &pb.HandlerDefinition{Subtype: subtype}
+		for _, m := range models {
+			handler.Models = append(handler.Models, m.String())
+		}
+		pMap.Handlers = append(pMap.Handlers, handler)
+	}
+	return pMap
+}
+
+// NewHandlerMapFromProto converts protobuf to HandlerMap.
+func NewHandlerMapFromProto(ctx context.Context, pMap *pb.HandlerMap, conn *grpc.ClientConn) (HandlerMap, error) {
+	hMap := make(HandlerMap)
+	refClient := grpcreflect.NewClient(ctx, reflectpb.NewServerReflectionClient(conn))
+	defer refClient.Reset()
+	reflSource := grpcurl.DescriptorSourceFromServer(ctx, refClient)
+
+	var errs error
+	for _, h := range pMap.GetHandlers() {
+		api := protoutils.ResourceNameFromProto(h.Subtype.Subtype).Subtype
+
+		symDesc, err := reflSource.FindSymbol(h.Subtype.ProtoService)
+		if err != nil {
+			errs = multierr.Combine(errs, err)
+			continue
+		}
+		svcDesc, ok := symDesc.(*desc.ServiceDescriptor)
+		if !ok {
+			return nil, errors.Errorf("expected descriptor to be service descriptor but got %T", symDesc)
+		}
+		subtype := &resource.RPCSubtype{
+			Subtype: api,
+			Desc:    svcDesc,
+		}
+		for _, m := range h.Models {
+			model, err := resource.NewModelFromString(m)
+			if err != nil {
+				return nil, err
+			}
+			hMap[*subtype] = append(hMap[*subtype], model)
+		}
+	}
+	return hMap, errs
+}
+
+// Module represents an external resource module that services components/services.
+type Module struct {
+	parent                  *client.RobotClient
+	grpcServer              *grpc.Server
+	logger                  *zap.SugaredLogger
+	mu                      sync.Mutex
+	ready                   bool
+	addr                    string
+	parentAddr              string
+	activeBackgroundWorkers sync.WaitGroup
+	handlers                HandlerMap
+	services                map[resource.Subtype]subtype.Service
+	pb.UnimplementedModuleServiceServer
+}
+
+// NewModule returns the basic module framework/structure.
+func NewModule(address string, logger *zap.SugaredLogger) *Module {
+	m := &Module{
+		logger:     logger,
+		addr:       address,
+		grpcServer: grpc.NewServer(),
+		ready:      true,
+		handlers:   make(HandlerMap),
+		services:   make(map[resource.Subtype]subtype.Service),
+	}
+	m.buildHandlerMap()
+	pb.RegisterModuleServiceServer(m.grpcServer, m)
+	reflection.Register(m.grpcServer)
+	return m
+}
+
+// Start starts the module service and grpc server.
+func (m *Module) Start(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	err := m.initSubtypeServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	oldMask := syscall.Umask(0o077)
```

Comment:
if this alleviates my security concerns, can you write a detailed comment about the security assumptions here?

Reply:
Replied to two other security notes, and I think this is the third. Should hopefully make sense in aggregate, but I'll make a details statement here as requested.

The basic security assumption is that the socket (effectively a file) doesn't need to be any more secure than our config files (which contain the security keys.) E.g. as long as we set the permissions to only allow access to the owner (e.g. the user running the process) we can trust that the unix file security model is sufficient. Adding additional security here (e.g. encryption) is unneeded, since any attack that would work here would work to get the secrets from the config file as well (and thus allow decryption of everything anyway.)

tl;dr nothing to do right here I believe, but should add an ownership check before dialing (both in parent and child) to be extra cautious, and set the umask BEFORE directory creation in web.StartLite() (to prevent a race attack creating a false socket.)